internal/mapdecode/decode_test.go

```diff
@@ -234,3 +237,234 @@ func TestDecode(t *testing.T) {
 		})
 	}
 }
+
+func TestFieldHook(t *testing.T) {
+	type myStruct struct {
+		SomeInt          int
+		SomeString       string
+		PtrToPtrToString **string
+
+		YAMLField string `yaml:"yamlKey"`
+
+		unexportedField string
+	}
+
+	typeOfInt := reflect.TypeOf(0)
+	typeOfString := reflect.TypeOf("hi")
+	typeOfPtrPtrString := reflect.PtrTo(reflect.PtrTo(typeOfString))
+
+	tests := []struct {
+		desc      string
+		setupHook func(*mockFieldHook)
+		give      interface{}
+		giveOpts  []Option // options besides FieldHook
+
+		want       myStruct
+		wantErrors []string
+	}{
+		{
+			desc: "not a map",
+			give: []interface{}{
+				map[string]interface{}{"a": 1},
+				map[string]interface{}{"b": 2},
+			},
+			wantErrors: []string{"expected a map, got 'slice'"},
+		},
+		{
+			desc:       "wrong map key",
+			give:       map[int]interface{}{1: "a"},
+			wantErrors: []string{"needs a map with string keys, has 'int' keys"},
+		},
+		{
+			desc: "string key",
+			give: map[string]interface{}{},
+		},
+		{
+			desc: "interface{} key",
+			give: map[interface{}]interface{}{},
+		},
+		{
+			desc: "unexported field",
+			give: map[string]string{"unexportedField": "hi"},
+		},
+		{
+			desc: "in-place updates",
+			give: map[string]interface{}{
+				"someInt":          1,
+				"PtrToPtrToString": "hello",
+			},
+			setupHook: func(h *mockFieldHook) {
+				h.Expect(_typeOfEmptyInterface, structField{
+					Name: "SomeInt",
+					Type: typeOfInt,
+				}, reflectEq{1}).Return(valueOf(42), nil)
+
+				h.Expect(_typeOfEmptyInterface, structField{
+					Name: "PtrToPtrToString",
+					Type: typeOfPtrPtrString,
+				}, reflectEq{"hello"}).Return(valueOf("world"), nil)
+			},
+			want: myStruct{
+				SomeInt:          42,
+				PtrToPtrToString: ptrToPtrToString("world"),
+			},
+		},
+		{
+			desc:     "field name override",
+			give:     map[string]interface{}{"yamlKey": "foo"},
```

Comment:
if we're going to keep case-insensitive searches, should we have a test case that has the wrong case